remote/remote_state.go

```diff
@@ -79,28 +81,46 @@ func (remoteState *RemoteState) NeedsInit(terragruntOptions *options.TerragruntO
 	if initializer, hasInitializer := remoteStateInitializers[remoteState.Backend]; hasInitializer {
 		// Remote state initializer says initialization is necessary
 		return initializer.NeedsInitialization(remoteState.Config, state.Backend, terragruntOptions)
-	} else if state.IsRemote() && remoteState.differsFrom(state.Backend, terragruntOptions) {
+	} else if state.IsRemote() {
+		needsInit, err := remoteState.differsFrom(state.Backend, terragruntOptions)
+		if nil != err {
+			return false, err
+		}
 		// If there's no remote state initializer, then just compare the the config values
-		return true, nil
+		return needsInit, nil
 	}
 
 	return false, nil
 }
 
 // Returns true if this remote state is different than the given remote state that is currently being used by terraform.
-func (remoteState *RemoteState) differsFrom(existingBackend *TerraformBackend, terragruntOptions *options.TerragruntOptions) bool {
+func (remoteState *RemoteState) differsFrom(existingBackend *TerraformBackend, terragruntOptions *options.TerragruntOptions) (bool, error) {
 	if existingBackend.Type != remoteState.Backend {
 		terragruntOptions.Logger.Printf("Backend type has changed from %s to %s", existingBackend.Type, remoteState.Backend)
-		return true
+		return true, nil
 	}
 
-	if !reflect.DeepEqual(existingBackend.Config, remoteState.Config) {
-		terragruntOptions.Logger.Printf("Backend config has changed from %s to %s", existingBackend.Config, remoteState.Config)
-		return true
+	var pConfig = make(map[string]interface{}, len(remoteState.Config))
+	include := terragruntOptions.GetIncludeConfig()
+	for key, val := range remoteState.Config {
+		if szval, ok := val.(string); ok {
+			pVal, err := shell.ResolveTerragruntConfigString(szval, include, terragruntOptions)
```

Comment:
Processing all interpolations in the `differsFrom` method feels like a big abstraction leak to me. I think we need to process them somewhere earlier, store the processed ones in `terragruntOptions`, and all other methods should not be required to know anything about interpolations.